packages/next/next-server/lib/router/utils/route-regex.ts

```diff
@@ -20,6 +20,10 @@ export function getRouteRegex(
   const parameterizedRoute = escapedRoute.replace(
     /\/\\\[([^/]+?)\\\](?=\/|$)/g,
     (_, $1) => {
+      const isOptional = /^\\\[.*\\\]$/.test($1)
```

Comment:
This should only support catch-all routes.

Reply:
@Timer As an alternative, would [failing](https://github.com/zeit/next.js/pull/12887/files#diff-6a1ab8ce5323e3179038754e716d22baR135) on usage of optional non-catch parameters be an acceptable solution?